reedstrm/baking_workflow_and_errors.md

```diff
@@ -0,0 +1,120 @@
+# Baking and Recipes
+
+Baking has become a critical step for publishing - however, it is not planned for it to cause publishing to fail. This is partly due to how long baking may take - much like PDF generation, it can take several hours, if there are external parts, like exercises, that need to be fetched and incorporated into the pages.
+
+This slightly odd situation (is baking a publish blocker, or not?) comes about partly due to derived copies - publishing a new version of content impacts not only the immediate book, but any book that shares pages with it. It was decided that failure of related books to bake is _not_ a publication blocker, so by extension, it is not for any book. However, we want to be careful to not break existing books by publishing new content, but we cannot test every book that contains a given page.
+
+## Selection of Recipe and Fallback
+
+So, we will implement two ways to mitigate this potential problem. First, will we keep track of when and with what recipe a given book successfully baked. This will point to the actual recipe bits - the file used. In order to use a given recipe for more than one book, we are extending the `print_style` attribute to also select a default recipe. Note that this can be updated independently of re-baking a given book.
+
+When a new version of a book is published, we will attempt to bake it with the recipe specified by the `print_style` parameter. There will be a table of recipes that are the _current_ recipe for given print style. If baking with that recipe fails, we will then retry baking with the last successful recipe for this book. If that succeeds, we update the webview, but also mark the state as `stale_recipe`. This way, the new content is available, even if changes in presentation from a new recipe are not. Note that uploading a new default recipe for a given `print_style` will also put books in this state.
+
+If the `print_style` parameter value does not have a default recipe associated with it, we will then check to see if a file with that name has been attached to the book. If there is no file, the book will be marked `Done/Success`. This will allow us to rollout recipes gradually for existing print styles. If there is a file, this will serve as a local recipe for baking a specific book, while removing the need for a "magic filename" from our processing code. The fallback described above would still be used - if the current book doesn't bake, use the most recent successful recipe, unless it is the same one we just tried.
+
+## Content Fallback and Delayed Baking
+
+If both attempts to bake fail, we leave the existing baked content in place, and mark the book as `stale_content`. The latest version of the content will still be available, by viewing the raw version, via the `?is_baked=false` query parameter.
```

Comment:
I think the 2 attempts are to use the _new_ recipe and if that fails, use the _old_ recipe that worked previously. I'm not terribly comfortable with having 2 attempts because then it is unclear _which_ attempt succeeded.

As an example, let's say QA/CM uploads a new recipe:

1. the book is correctly marked as pending
2. the book is correctly marked as baked
3. QA/CM checks that the shiny new feature is in the book
4. QA/CM scratches head because that new feature isn't in the book

Oh wait, maybe the book would have a ⚠️  instead of a ✅  when baking using the old recipe. If so, then I'm 🆗  but it should probably be clarified here what happens.

Maybe something like "If both attempts to bake (fallback to previous recipe, with a ⚠️ on success) fail, we leave..."